python_bindings/correctness/pystub.py

```diff
@@ -174,10 +174,27 @@ def test_complexstub():
                     simple_input=input,
                     array_input=[ input, input ],
                     float_arg=float_arg,
-                    int_arg=[ int_arg, int_arg ],
-                    untyped_buffer_output_type="uint8",
+                    int_arg=[int_arg, int_arg],
+                    # We can put GeneratorParams anywhere in the list we want;
+                    # they will be examined and applied before any inputs are
+                    # processed, and thus, any type/dimension constraints will be
+                    # enforced on inputs.
+                    #
+                    # Note that in Python, synthetic GeneratorParams
+                    # replace the '.' with a double underscore ('__'), so setting
+                    # untyped_buffer_output.type is accomplished like so:
```

Comment:
Why is this the case? Can't it be something like `untyped_buffer_output={"type": "uint8"}`?

Reply:
eh... I guess it could be? I avoided that mainly because when I wrote this I thought I'd want to implement the same thing for C++ stubs, in which case having a common syntax would be desirable, and the __ approach is more straightforward to manage for C++. I'll give your suggestion a try and see how it feels.

Reply:
Actually, there *is* a very good reason that syntax won't work as-is: inputs that also need their type (etc) set. Currently:

    untyped_buffer_input=constant_image,  // line 173 in this PR
    untyped_buffer_input__type="uint8",  // line 186 in this PR

With this proposal, we'd have the obviously-illegal:

    untyped_buffer_input=constant_image, 
    untyped_buffer_input={"type": "uint8"},

We could finesse this in various ways, of course, e.g.

    # require a dict that includes value as well as type, dim, etc for these cases
    untyped_buffer_input={"value": constant_image, "type": "uint8"},

...but that feels weird and awkward IMHO.

Reply:
I see your point and I keep going back and forth on whether using `__` or a dictionary like this is _more_ weird and awkward. Double underscores generally indicate language-level magic in Python, which isn't really the case here. On the other hand, the `"value"` key isn't really mirrored anywhere else in the API.

Why can't we get the type out of the `constant_image` object? When it's created on line 150, it's typed with `hl.UInt(8)`.

Reply:
> Why can't we get the type out of the constant_image object?

(sorta-topic-change to other work going on elsewhere...)

We *can*, the question is whether we *should*: there is a distinction between "is passed-in-argument well-typed" (yes, it is), vs "is the type required by the receiver well-defined or not" (in this case, not).

Currently, in C++ Generators (and by extension, the almost-never-used C++-based stubs), we take the approach that we don't require the receiver to be explicitly typed iff we can infer a concrete type from the passed-in-argument. This seems appealing, and works, but has several drawbacks that I'd like to avoid in the still-work-in-progress Python generator design:
- It complicates the internal code for Generators, since we may no longer know the type constraints on Inputs/Outputs until you actually execute the Generator to produce the IR.
- It impairs the readability for Generators, since you may not be able to look at the code (or even the code + the build file containing GeneratorParam values) and know what the expected types are. (For 'types', also include 'dimensions' and 'array-size', of course.)

IMHO a superior design would have been to require that all 'untyped' things be explicitly specified via GeneratorParams, so that no Generator code must be executed to deduce the signature of the resulting function. Trying to change that design point for C++ Generators would be a fool's errand at this point, but I'd really like to make this a design point to enforce in Python.

Reply:
> I see your point and I keep going back and forth on whether using `__` or a dictionary like this is _more_ weird and awkward. Double underscores generally indicate language-level magic in Python, which isn't really the case here. 

Yeah, I see your point and don't disagree. 

Arguably, the 'bug' here is that 'synthetic' params like this were created with a '.' in their names as a way to ensure they didn't overlap with any other param, since '.' was not and is not a valid user-provided name -- since there wasn't (and still isn't) a way to specify these for C++ stubs (which are rarely used, a nightmare to implement and use, and probably will never be enhanced), this isn't a limitation, but for Python-based stubs (which are potentially *much* more useful, for various reasons), being able to support these would be quite useful.

So the real question here is: is there a syntax that fits reasonably with Python named-argument convention which is reasonable?

Reply:
Any further thoughts on this? I'm not especially happy with the `__` approach, but lack a better idea at the moment...